source/initial_composition/slab_model.cc

```diff
@@ -0,0 +1,123 @@
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+ */
+
+
+#include <aspect/global.h>
+#include <aspect/initial_composition/slab_model.h>
+#include <aspect/utilities.h>
+
+
+namespace aspect
+{
+  namespace InitialComposition
+  {
+    template <int dim>
+    void
+    SlabModel<dim>::initialize ()
+    {
+      // The input slabs are defined from the surface of the model
+      std::set<types::boundary_id> surface_boundary_set;
+      surface_boundary_set.insert(this->get_geometry_model().translate_symbolic_boundary_name_to_id("top"));
+
+      // The two columns correspond to slabs depth and thickness
+      slab_boundary.initialize(surface_boundary_set, 2);
+    }
+
+
+    template <int dim>
+    double
+    SlabModel<dim>::
+    initial_composition (const Point<dim> &position,
+                         const unsigned int compositional_index) const
+    {
+      const unsigned int surface_boundary_id = this->get_geometry_model().translate_symbolic_boundary_name_to_id("top");
+      const unsigned int slab_index          = this->introspection().compositional_index_for_name("slabs");
+
+      const double depth                     = this->get_geometry_model().depth(position);
+      double slab_composition                = 0;
+
+      // The first column after the points correspond to the slabs depth and the second to the slabs thickness
+      double slab_depths    = slab_boundary.get_data_component(surface_boundary_id, position, 0);
+      double slab_thickness = slab_boundary.get_data_component(surface_boundary_id, position, 1);
+
+      // The input ascii file is structure and as a result has many data points where slabs are absent. We give a very high
+      // number at those locations and therefore the first two conditions check if we are within the slabs.
+      // In the input file, slab depths are to the top of the slab surface.
+      // The hyperbolic tangent function smooths the slabs in the depth direction.
+      if ( (compositional_index == slab_index) && (slab_depths < 1e10) && (slab_thickness < 1e10) &&
+           (depth >= slab_depths) && (depth <= slab_depths + slab_thickness) )
+        {
+          const double slab_center    = slab_depths + slab_thickness/2.;
+          const double half_thickness = slab_thickness/2;
+          slab_composition            = ( 1 - std::tanh( 10 * (std::abs (slab_center - depth) - half_thickness)/half_thickness ) )/2;
```

Comment:
At the moment you have hard-coded the thickness of the transition region and let it scale with the thickness of the slab. Is that  intended? I would have thought it is more useful to either have a constant transition region that is an input parameter (with the disadvantage like we discussed that it is only diffused in depth, not laterally). The alternative would be to simply not have a transition region at all, which may be simpler. E.g. something like this:
```suggestion
          slab_composition = 1.0;
```